iml-services/iml-device/src/virtual_device/mod.rs

```diff
@@ -0,0 +1,292 @@
+        let device_to_insert = NewChromaCoreDevice {
+            fqdn: f.to_string(),
+            devices: serde_json::to_value(d).expect("Could not convert other Device to JSON."),
+        };
+
+        let new_device = diesel::insert_into(table)
+            .values(device_to_insert)
+            .on_conflict(fqdn)
+            .do_update()
+            .set(chroma_core_device::devices.eq(excluded(chroma_core_device::devices)))
+            .get_result_async::<ChromaCoreDevice>(pool)
+            .await
+            .expect("Error saving new device");
+
+        tracing::info!("Inserted devices from host {}", new_device.fqdn);
+        tracing::trace!("Inserted devices {:?}", new_device);
+    }
+}
+
+pub async fn get_other_devices(f: &Fqdn, pool: &DbPool) -> Vec<(Fqdn, Device)> {
+    let other_devices = table
+        .filter(fqdn.ne(f.to_string()))
+        .load_async::<ChromaCoreDevice>(&pool)
+        .await
+        .expect("Error getting devices from other hosts");
+
+    other_devices
+        .into_iter()
+        .map(|d| {
+            (
+                Fqdn(d.fqdn),
+                serde_json::from_value(d.devices)
+                    .expect("Couldn't deserialize Device from JSON when reading from DB"),
+            )
+        })
+        .collect()
+}
+
+pub fn update_virtual_devices(devices: Vec<(Fqdn, Device)>) -> Vec<(Fqdn, Device)> {
+    let mut results = vec![];
+
+    // As there are multipath devices and we don't compare major, minor, devpath, paths fields,
+    // there will be duplicates, of which we'll end up using only one,
+    // since when inserting we again search using same fields.
+    // So keep parents in a set to avoid iterating the same devices in insert_virtual_devices twice.
+    let mut parents = collections::HashSet::new();
+    let devices2 = devices.clone();
+
+    let len = devices.len();
+    for (i, (f, d)) in devices.into_iter().enumerate() {
+        let ps = collect_virtual_device_parents(&d, 0, None);
+        // The incoming tree (from current host) can have less virtual device parents, than trees from the DB (from other hosts).
+        // In the incoming data there are only virtual devices that are local to that host (i.e. are mounted there).
+        // In the database, there are virtual devices that are collected from all of hosts.
+        // So the note is to reflect that. We push the incoming data to the end of the `Vec` so it's last.
+        let note = if i == len - 1 { " (incoming)" } else { "" };
+        tracing::info!(
+            "Collected {:3} parents at {:25} host{}",
+            ps.len(),
+            f.to_string(),
+            note
+        );
+        parents.extend(ps);
+    }
+
+    for (f, d) in devices2 {
+        let dd = insert_virtual_devices(d, &parents);
+
+        results.push((f, dd));
+    }
+
+    results
+}
+
+fn collect_virtual_device_parents(
+    d: &Device,
+    level: usize,
+    parent: Option<&Device>,
+) -> Vec<Device> {
```

Comment:
Why can't we use serials here instead of collecting the whole parent device?

Reply:
We currently compare not only serials as I mention in another comment, and we need children of this device.
I now return a reference instead of cloning.